### PR TITLE
refactor(base): clean up change indicators

### PR DESCRIPTION
**Pull request by mariuslundgard**
### Description

- Fixes an issue that caused change indicator bars not showing on mobile
- Improves types
- Refactors for readability
- Performance improvements

**Bonus**: the change indicators/connectors can now be tested in the workshop:
https://studio-workshop-git-refactor-base-change-indicators.sanity.build/base/change-indicators/change-connector

### CODE REVIEW

File: packages/@sanity/base/src/change-indicators/ChangeIndicator.tsx [hunk 1]
```diff
@@ -206,7 +205,15 @@ export const ChangeIndicatorWithProvidedFullPath = ({
...
-}: any) => {
+}: {
+  className?: string
+  disabled?: boolean
+  path: Path
+  value: unknown
+  hasFocus: boolean
+  compareDeep?: boolean
+  children?: React.ReactNode
+}) => {
```
Comment (ricokahler) [inline packages/@sanity/base/src/change-indicators/ChangeIndicator.tsx:216]: Might be better as `ChangeIndicatorWithProvidedFullPathProps` 🤷

File: packages/@sanity/base/src/change-indicators/helpers/getOffsetsTo.ts [hunk 2]
```diff
@@ -25,33 +33,15 @@ const getOffsetsTo = (source: HTMLElement, target: HTMLElement) => {
...
-    console.time(label)
-    const ret = fn(...args)
-    // eslint-disable-next-line no-console
-    console.timeEnd(label)
-    return ret
+    el = el.offsetParent as HTMLElement
   }
-}
 
-export default getOffsetsTo
```
Comment (ricokahler) [inline packages/@sanity/base/src/change-indicators/helpers/getOffsetsTo.ts:57]: we gotta add [this one](https://github.com/import-js/eslint-plugin-import/blob/7c382f02a6cafe676ee754a896092040dad8daf6/docs/rules/no-default-export.md) eventually

File: packages/@sanity/base/src/change-indicators/overlay/ChangeConnectorRoot.tsx [hunk 3]
```diff
@@ -21,8 +19,8 @@ function EnabledChangeConnectorRoot({
   onSetFocus,
   isReviewChangesOpen,
   onOpenReviewChanges,
-}: Props) {
-  const [rootRef, setRootRef] = React.useState<HTMLDivElement | null>()
+}: EnabledChangeConnectorRootProps) {
+  const [rootElement, setRootRef] = React.useState<HTMLDivElement | null>()
```
Comment (ricokahler) [inline packages/@sanity/base/src/change-indicators/overlay/ChangeConnectorRoot.tsx:23]:
```diff
-  const [rootElement, setRootRef] = React.useState<HTMLDivElement | null>()
+  const [rootElement, setRootElement] = React.useState<HTMLDivElement | null>()
```

File: packages/@sanity/base/src/change-indicators/overlay/useResizeObserver.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import {useEffect} from 'react'
+import {resizeObserver} from '../../util/resizeObserver'
+
+export function useResizeObserver(
+  element: HTMLDivElement,
+  onResize: (event: ResizeObserverEntry) => void
+): void {
+  useEffect(() => resizeObserver.observe(element, onResize), [element, onResize])
```
Comment (ricokahler) [inline packages/@sanity/base/src/change-indicators/overlay/useResizeObserver.ts:8]: old code but i think this is missing a clean up function to call `resizeObserver.disconnect`

File: packages/@sanity/base/src/change-indicators/overlay/ConnectorsOverlay.tsx [hunk 5]
```diff
@@ -81,50 +73,119 @@ export const ConnectorsOverlay = React.memo(function ConnectorsOverlay(props: Pr
...
+  )
+
+  const visibleConnectors = useMemo(
+    () => sortBy(connectors, (c) => -c.field.path.length).slice(0, 1),
+    [connectors]
+  )
+
+  const handleScrollOrResize = useCallback(() => {
+    setState(getState(allReportedValues, hovered, byId, rootElement))
+  }, [byId, allReportedValues, hovered, rootElement])
```
Comment (ricokahler) [inline packages/@sanity/base/src/change-indicators/overlay/ConnectorsOverlay.tsx:111]: very nice refactor 💯

File: packages/@sanity/base/src/change-indicators/overlay/ConnectorsOverlay.tsx [hunk 6]
```diff
@@ -81,50 +73,119 @@ export const ConnectorsOverlay = React.memo(function ConnectorsOverlay(props: Pr
...
+function ConnectorGroup(props: {
+  field: TrackedChange & {id: string; rect: Rect; bounds: Rect}
+  change: TrackedChange & {id: string; rect: Rect; bounds: Rect}
+  hasFocus: boolean
+  hasHover: boolean
+  hasRevertHover: boolean
+  setHovered: (id: string | null) => void
+  onSetFocus: (nextFocusPath: Path) => void
+  isHoverConnector: boolean
+}) {
```
Comment (ricokahler) [inline packages/@sanity/base/src/change-indicators/overlay/ConnectorsOverlay.tsx:154]: `ConnectorGroupProps`? 🤷

**Review by ricokahler — APPROVED** [review]
I had one comment around the resize observer that i think is worth looking into but besides that, this is a solid refactor 😄

File: packages/@sanity/base/src/change-indicators/overlay/useResizeObserver.ts (hunk 4 above)
Comment (mariuslundgard) [reply packages/@sanity/base/src/change-indicators/overlay/useResizeObserver.ts:8]: The `resizeObserver` does that internally in this case.

**Pull request merged**